src/Deserialization.cpp

```diff
@@ -0,0 +1,1358 @@
+        const Serialize::Variable *variable_expr = (const Serialize::Variable *)expr;
+        auto name = deserialize_string(variable_expr->name());
+        auto type = deserialize_type(variable_expr->type());
+        auto param_name = deserialize_string(variable_expr->param_name());
+        Parameter param;
+        if (auto it = non_serialized_parameters.find(param_name); it != non_serialized_parameters.end()) {
+            param = it->second;
+        } else if (auto it = parameters_in_pipeline.find(param_name); it != parameters_in_pipeline.end()) {
+            param = it->second;
+        }
+        Buffer<> image;
+        auto image_name = deserialize_string(variable_expr->image_name());
+        if (auto it = buffers_in_pipeline.find(image_name); it != buffers_in_pipeline.end()) {
+            image = it->second;
+        }
+        auto reduction_domain = deserialize_reduction_domain(variable_expr->reduction_domain());
+        return Variable::make(type, name, image, param, reduction_domain);
+    }
+    case Serialize::Expr::Expr_Shuffle: {
+        const Serialize::Shuffle *shuffle_expr = (const Serialize::Shuffle *)expr;
+        std::vector<Expr> vectors = deserialize_expr_vector(shuffle_expr->vectors_type(), shuffle_expr->vectors());
+        auto indices_serialized = shuffle_expr->indices();
+        std::vector<int32_t> indices;
+        indices.reserve(indices_serialized->size());
+        for (size_t i = 0; i < indices_serialized->size(); ++i) {
+            indices.push_back(indices_serialized->Get(i));
+        }
+        return Shuffle::make(vectors, indices);
+    }
+    case Serialize::Expr::Expr_VectorReduce: {
+        const Serialize::VectorReduce *vector_reduce_expr = (const Serialize::VectorReduce *)expr;
+        auto value = deserialize_expr(vector_reduce_expr->value_type(), vector_reduce_expr->value());
+        auto reduction_op = deserialize_vector_reduce_op(vector_reduce_expr->reduction_op());
+        int32_t lanes = vector_reduce_expr->lanes();
+        return VectorReduce::make(reduction_op, value, lanes);
+    }
+    case Serialize::Expr::Expr_UndefinedExpr: {
+        return Expr();
+    }
+    default: {
+        user_assert(false) << "unknown type code " << type_code << "\n";
+        return Expr();
+    }
+    }
+}
+
+std::vector<Expr> Deserializer::deserialize_expr_vector(const flatbuffers::Vector<uint8_t> *exprs_types, const flatbuffers::Vector<flatbuffers::Offset<void>> *exprs_serialized) {
+    assert(exprs_types != nullptr);
+    assert(exprs_serialized != nullptr);
+    std::vector<Expr> result;
+    result.reserve(exprs_serialized->size());
+    for (size_t i = 0; i < exprs_serialized->size(); ++i) {
+        auto expr = deserialize_expr(static_cast<Serialize::Expr>(exprs_types->Get(i)), exprs_serialized->Get(i));
+        result.push_back(expr);
+    }
+    return result;
+}
+
+Range Deserializer::deserialize_range(const Serialize::Range *range) {
+    assert(range != nullptr);
```

Comment:
Some of the null checks in this file use `user_assert` (which always checks) and some use `assert` (which depends on compiler mode), is this intentional?

Reply:
no, we should stick to `user_assert`